### ISSUE TITLE
Discussion about Farmer future: Move from ARM templates to AZ CLI

### ISSUE THREAD

**Issue by Dzoukr** [CompositionalIT/farmer#283]
Hello @isaacabraham,

since our morning discussion about how awesome is to have combination of ARM templates with Post-deploy tasks, I am thinking about this:

I love what Farmer can do right now. For example the thing I can _setup_ infrastructure and _deploy_ to it is amazing. The problem here is, that to have both, you need to combine ARM with AZ CLI and as a user of Farmer you have to be _aware of it_ - I need to know that if I'll only export JSON, my app won't be deployed.

```fsharp
let myApp = functions {
    name "farmer-amazing-function-app" // <-- ARM
    use_extension_version V3 // <-- ARM
    zip_deploy "path/to/my/app" // <-- AZ CLI
}
```

So here it comes. What if we move away from ARM to AZ CLI completely? Yes, you have to have it installed (or supported on CI server), but:

- You don't have to know what is part of ARM and what is part of post-deploy task
- You can support way more features currently not available in ARM
- You can use well documented AZ CLI instead of terribly semi-documented ARM
- There is no breaking change in API - library is greatly designed so Builders and their interpretation are totally separated

ARM is about setup of _dead_ infrastructure, but over AZ CLI you can make it alive, which is a kickass feature. I think this would leverage Farmer to completely new level.

What do you think about it? Please take it as trigger for discussion about future and feel free to close immediatelly if you found it completely silly.

**Comment by isaacabraham** [comment]
Hah :-) I think this is not a terrible idea but I suspect it would need to be an alternative to the ARM template generation as opposed to an instead of. Definitely not a v1 feature but it's good to keep this around - especially as you mention, there are some nice things the CLI gives us that ARM doesn't. HOWEVER we would also lose the idempotent "guarantees" (more or less...) that ARM gives us - unless we created our own state service to track deployments...

**Comment by Dzoukr** [comment]
Good points (from both of you). I am glad the discussion flows. I like the alternative approach, btw...

**Comment by ctolkien** [comment]
>You don't have to know what is part of ARM and what is part of post-deploy task

I feel like this could be resolved via making it clearer just via the API? `CLI.zip_deploy` (pseudocode)

**Comment by isaacabraham** [comment]
Closed in favour of #483
